docs/design/graceful-domain-failover/3051-graceful-domain-failover.md

```diff
@@ -0,0 +1,239 @@
+# Design doc: Cadence graceful domain failover
+
+Author: Cadence Team
+
+Last updated: Mar 2020
+
+Reference: [#3051](https://github.com/uber/cadence/issues/3051)
+
+
+## Abstract
+
+Cadence supports domain failover with multi-cluster setup. However, the problems with the current failover are:
+1. Workflow progress could be lost.
+2. No causal consistency guarantee on domain level.
+
+The graceful domain failover uses to solve those two problems.
+
+## Use cases
+
+Users can trigger graceful domain failover via Cadence CLI (timeout is optional):
+
+`cadence --domain cadence-global domain update -failover_type graceful -active_cluster XYZ -failover_timeout 120s`
+
+Users can force complete domain failover via Cadence CLI:
+
+`cadence --domain cadence-global domain update -failover_type force -active_cluster XYZ`
+
+## Prerequisites
+There are conditions before starting a graceful domain failover.
+1. All clusters are available.
+2. The networking between clusters are available.
+3. The domain state has to be stable prior to graceful failover.
+
+## Limitation
+No concurrent graceful domain failover will be supported.
```

Comment:
Do we actually reject requests for concurrent graceful failovers or just provide no guarantees if the user issues concurrent graceful failovers?

Reply:
In race condition, users could issue concurrent graceful failover and if this happens then we don't guarantee anything